eth/stagedsync/stage_hashstate.go

```diff
@@ -169,34 +151,126 @@ func PromoteHashedStateCleanly(logPrefix string, db kv.RwTx, cfg HashStateCfg, q
 	)
 }
 
-func keyTransformExtractFunc(transformKey func([]byte) ([]byte, error)) etl.ExtractFunc {
-	return func(k, v []byte, next etl.ExtractNextFunc) error {
-		newK, err := transformKey(k)
+func readPlainStateOnce(
+	logPrefix string,
+	db kv.RwTx,
+	fromBucket string,
+	tmpdir string,
+	loadFunc etl.LoadFunc,
+	quit <-chan struct{},
+) error {
+	bufferSize := etl.BufferOptimalSize
+	buffer := etl.NewSortableBuffer(bufferSize)
+
+	collector1 := etl.NewCollector(logPrefix, tmpdir, buffer)
+	collector2 := etl.NewCollector(logPrefix, tmpdir, buffer)
+	defer func() {
+		collector1.Close()
+		collector2.Close()
+	}()
+
+	t := time.Now()
+	logEvery := time.NewTicker(30 * time.Second)
+	defer logEvery.Stop()
+	var m runtime.MemStats
+
+	c, err := db.Cursor(fromBucket)
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	convertAccFunc := func(key []byte) ([]byte, error) {
+		hash, err := common.HashData(key)
+		return hash[:], err
+	}
+
+	convertStorageFunc := func(key []byte) ([]byte, error) {
+		addrHash, err := common.HashData(key[:length.Addr])
 		if err != nil {
-			return err
+			return nil, err
 		}
-		return next(k, newK, v)
+		inc := binary.BigEndian.Uint64(key[length.Addr:])
+		secKey, err := common.HashData(key[length.Addr+length.Incarnation:])
+		if err != nil {
+			return nil, err
+		}
+		compositeKey := dbutils.GenerateCompositeStorageKey(addrHash, inc, secKey)
+		return compositeKey, nil
 	}
-}
 
-func keyTransformExtractAcc(transformKey func([]byte) ([]byte, error)) etl.ExtractFunc {
-	return func(k, v []byte, next etl.ExtractNextFunc) error {
-		if len(k) != 20 {
-			return nil
+	var startkey, endkey []byte
+
+	// reading kv.PlainState
+	for k, v, e := c.Seek(startkey); k != nil; k, v, e = c.Next() {
+		if e != nil {
+			return e
 		}
-		newK, err := transformKey(k)
-		if err != nil {
+		if err := libcommon.Stopped(quit); err != nil {
 			return err
 		}
-		return next(k, newK, v)
-	}
-}
 
-func keyTransformExtractStorage(transformKey func([]byte) ([]byte, error)) etl.ExtractFunc {
-	return func(k, v []byte, next etl.ExtractNextFunc) error {
 		if len(k) == 20 {
+			newK, err := convertAccFunc(k)
+			if err != nil {
+				return err
+			}
+			if err := collector1.Collect(newK, v); err != nil {
+				return err
+			}
+		} else {
+			newK, err := convertStorageFunc(k)
+			if err != nil {
+				return err
+			}
+			if err := collector2.Collect(newK, v); err != nil {
+				return err
+			}
+		}
+
+		select {
+		default:
+		case <-logEvery.C:
+			logArs := []interface{}{"from", fromBucket}
+			var additionalLogArguments etl.AdditionalLogArguments
+			if additionalLogArguments != nil {
```

Comment:
checking for nil of variable which just created.

Reply:
Do we actually have to declare `var additionalLogArguments etl.AdditionalLogArguments`. I would replace it with this:
```
case <-logEvery.C:
			logArs := []interface{}{"from", fromBucket}
		        logArs = append(logArs, "current key", fmt.Sprintf("%x...", k[:6]))
```                                                                                                                                                                                
 what do you think?

Reply:
Nope. 1 liner is fine here: 
```
log.Info(fmt.Sprintf("[%s] ETL [1/2] Extracting", logPrefix), "current key", fmt.Sprintf("%x...", k[:6]), "alloc", libcommon.ByteCount(m.Alloc), "sys", libcommon.ByteCount(m.Sys))) 
```